pkg/syncer/cnsoperator/controller/triggercsifullsync/triggercsifullsync_controller.go

```diff
@@ -0,0 +1,342 @@
+var _ reconcile.Reconciler = &ReconcileTriggerCsiFullSync{}
+
+// ReconcileTriggerCsiFullSync reconciles a TriggerCsiFullSync object
+type ReconcileTriggerCsiFullSync struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client        client.Client
+	scheme        *runtime.Scheme
+	clusterFlavor cnstypes.CnsClusterFlavor
+	configInfo    *types.ConfigInfo
+	recorder      record.EventRecorder
+}
+
+// Reconcile reads that state of the cluster for a TriggerCsiFullSync object and makes changes based on the state read
+// and what is in the TriggerCsiFullSync.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileTriggerCsiFullSync) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx, log := logger.GetNewContextWithLogger()
+	// Fetch the TriggerCsiFullSync instance
+	instance := &triggercsifullsyncv1alpha1.TriggerCsiFullSync{}
+	err := r.client.Get(ctx, request.NamespacedName, instance)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Infof("TriggerCsiFullSync resource not found. Ignoring since object must be deleted.")
+			return reconcile.Result{}, nil
+		}
+		log.Errorf("Error reading the TriggerCsiFullSync with name: %q. Err: %+v",
+			request.Name, "", err)
+		// Error reading the object - return with err
+		return reconcile.Result{}, err
+	}
+
+	// Initialize backOffDuration for the instance, if required.
+	backOffDurationMapMutex.Lock()
+	var timeout time.Duration
+	if _, exists := backOffDuration[instance.Name]; !exists {
+		backOffDuration[instance.Name] = time.Second
+	}
+	timeout = backOffDuration[instance.Name]
+	backOffDurationMapMutex.Unlock()
+
+	// Ignore TriggerCsiFullSync instances other than csifullsync
+	if instance.Name != common.TriggerCsiFullSyncCRName {
+		msg := fmt.Sprintf("Only %q should be used to trigger full sync and not %q",
+			common.TriggerCsiFullSyncCRName, instance.Name)
+		log.Error(msg)
+		recordEvent(ctx, r, instance, v1.EventTypeWarning, msg)
+		return reconcile.Result{}, nil
+	}
+
+	// Ignore any updates on TriggerCsiFullSync instance with TriggerSyncID set to 0
+	if instance.Spec.TriggerSyncID == 0 || instance.Spec.TriggerSyncID == instance.Status.LastTriggerSyncID {
+		return reconcile.Result{}, nil
+	}
+
+	// If TriggerSyncID is not one greater than LastTriggerSyncID, raise an event that
+	// the trigger sync will be ignored
+	if instance.Spec.TriggerSyncID != instance.Status.LastTriggerSyncID+1 {
+		msg := fmt.Sprintf("TriggerSyncID: %d is invalid. TriggerSyncID should be one greater than LastTriggerSyncID.",
+			instance.Spec.TriggerSyncID)
+		log.Error(msg)
+		recordEvent(ctx, r, instance, v1.EventTypeWarning, msg)
+		return reconcile.Result{}, nil
+	}
+
```

Comment:
What will happen if the TriggerCsiFullSync API object is deleted while a sync is in-progress?

Reply:
TriggerCsiFullSync CR is created by our CSI driver as part of syncer init. Initial values will be 0 for triggersyncId. User can then update the CR from 0 to 1 to trigger full sync.